### PR TITLE
Migration of masked_image_modeling_with_autoencoders to keras 3

### PR DESCRIPTION
**Pull request by maitry63**
This PR migrates the Masked image modeling with Autoencoders tutorial from Keras 2 to Keras 3. The implementation is now fully backend-agnostic, allowing it to run seamlessly on JAX, PyTorch, and TensorFlow ([notebook](https://colab.sandbox.google.com/gist/maitry63/0f1ae8fc5d552177313553a3c9257e39/copy-of-masked_image_modeling.ipynb#scrollTo=p7pMLFYVphMH)).

### CODE REVIEW

**Comment by maitry63** [comment]
@gemini-code-assist review

**Comment by maitry63** [comment]
@gemini-code-assist review

**Comment by maitry63** [comment]
@gemini-code-assist review

File: examples/vision/masked_image_modeling.py [hunk 1]
```diff
@@ -45,22 +46,21 @@
...
-from keras import layers
-
 import matplotlib.pyplot as plt
 import numpy as np
-import random
+from keras import layers, ops
 
 # Setting seeds for reproducibility.
 SEED = 42
 keras.utils.set_random_seed(SEED)
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py:62]: Remove this extra line space.

File: examples/vision/masked_image_modeling.py [hunk 2]
```diff
@@ -82,7 +83,7 @@
 WEIGHT_DECAY = 1e-4
 
 # PRETRAINING
-EPOCHS = 100
+EPOCHS = 1
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py]: Can you keep the more epochs, if 100 epochs fits in the colab T4 instance, keep that.

File: examples/vision/masked_image_modeling.py [hunk 3]
```diff
@@ -405,12 +398,50 @@ def generate_masked_image(self, patches, unmask_indices):
...
 """
-Let's see the masking process in action on a sample image.
+## Prepare the Dataset
+"""
+train_ds = MaskedImageDataset(x_train, batch_size=BATCH_SIZE, shuffle=True)
+val_ds = MaskedImageDataset(x_val, batch_size=BATCH_SIZE)
+test_ds = MaskedImageDataset(x_test, batch_size=BATCH_SIZE)
+
 """
+Let's visualize the image patches.
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py]: Let's visualize the image patches. --> ### Let's visualize the image patches.

File: examples/vision/masked_image_modeling.py [hunk 4]
```diff
@@ -656,17 +697,15 @@ def test_step(self, images):
     decoder=decoder,
 )
 
+# Taking a batch of test inputs to measure model's progress.
+test_images = next(iter(test_ds))
+
 """
 ## Training callbacks
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py:703]: Add """ after this line and retain back the same docstring type as before, don't merge multiple headings or blocks in a single docstring.

File: examples/vision/masked_image_modeling.py [hunk 5]
```diff
@@ -852,38 +892,68 @@ def __call__(self, step):
 Another approach would be to use a learnable dummy token inside the encoder during
 pretraining (resembling the [CLS] token). Then we can extract representations from that
 token during the downstream tasks.
-"""
 
-"""
 ### Prepare datasets for linear probing
 """
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py:902]: These should be in a different doc string to differetiate different sections. Revert this back to previous state.

File: examples/vision/masked_image_modeling.py [hunk 6]
```diff
@@ -852,38 +892,68 @@ def __call__(self, step):
...
+train_ds = prepare_data(x_train, y_train, BATCH_SIZE, is_train=True)
+val_ds = prepare_data(x_val, y_val, BATCH_SIZE, is_train=False)
+test_ds = prepare_data(x_test, y_test, BATCH_SIZE, is_train=False)
 
 """
 ### Perform linear probing
 """
-
-linear_probe_epochs = 50
+linear_probe_epochs = 2
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py]: Increase the number fo epochs. 2 is too small.

File: examples/vision/masked_image_modeling.py [hunk 7]
```diff
@@ -914,9 +983,7 @@ def prepare_data(images, labels, is_train=True):
 in a fully supervised manner. This gave us ~76% test top-1 accuracy. The authors of
 MAE demonstrates strong performance on the ImageNet-1k dataset as well as
 other downstream tasks like object detection and semantic segmentation.
-"""
 
-"""
 ## Final notes
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py:995]: Retain the string structure to previous state.

File: examples/vision/masked_image_modeling.py [hunk 8]
```diff
@@ -803,15 +845,12 @@ def __call__(self, step):
...
+metrics = mae_model.evaluate(test_ds, return_dict=True)
+print(f"Loss: {metrics['loss']:.2f}")
 
 """
 ## Evaluation with linear probing
-"""
 
-"""
 ### Extract the encoder model along with other layers
 """
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py:859]: This should be same as before:

```
"""
## Evaluation with linear probing
"""

"""
### Extract the encoder model along with other layers
"""
```

File: examples/vision/masked_image_modeling.py [hunk 9]
```diff
@@ -656,17 +696,15 @@ def test_step(self, images):
     decoder=decoder,
 )
 
-"""
+# Taking a batch of test inputs to measure model's progress.
+test_images = next(iter(test_ds))
+
 ## Training callbacks
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py:703]: This should be 

```
"""
## Training callbacks
"""
```

File: examples/vision/masked_image_modeling.py [hunk 10]
```diff
@@ -907,36 +978,16 @@ def prepare_data(images, labels, is_train=True):
...
-in 2020 but we could not prevent the network from representation collapse, and
-hence we did not get strong downstream performance.
+Example available on HuggingFace
 
-We would like to thank [Xinlei Chen](http://xinleic.xyz/)
-(one of the authors of MAE) for helpful discussions. We are grateful to
-[JarvisLabs](https://jarvislabs.ai/) and
-[Google Developers Experts](https://developers.google.com/programs/experts/)
-program for helping with GPU credits.
+| Trained Model | Demo |
```
Comment (sachinprasadhs) [inline examples/vision/masked_image_modeling.py]: One example link is broken and other one is having keras 2 model weight file.
Instead of this, you can link the examples which are migrated to Keras 3 from the previous references like below:

```
- [SimCLR](https://keras.io/examples/vision/semisupervised_simclr/)
- [NNCLR](https://keras.io/examples/vision/nnclr)
- [SimSiam](https://keras.io/examples/vision/simsiam)
```

Comment (sachinprasadhs) [reply examples/vision/masked_image_modeling.py]: Also, you can retain the credits mentioned at the end.

**Comment by sachinprasadhs** [comment]
Thanks, Looks good. Please generate .md and .ipynb files.
